netbench-orchestrator/src/russula/mod.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{task::Poll, time::Duration};
+use std::{collections::BTreeSet, net::SocketAddr};
+use tokio::net::TcpStream;
+use tracing::{error, info};
+
+mod error;
+mod event;
+mod network_utils;
+mod protocol;
+mod states;
+
+use error::{RussulaError, RussulaResult};
+use protocol::Protocol;
+
+const CONNENT_RETRY_ATTEMPT: usize = 10;
+
+#[derive(Debug, Copy, Clone)]
+pub enum ProtocolState {
+    /// The endpoint has established connection with its peer and
+    /// is ready to make progress.
+    Ready,
+
+    /// Indicates the protocol's terminal state.
+    Done,
+
+    /// Indicates that worker are running and accepting work.
+    ///
+    /// For netbench this state be used to confirm that all servers are
+    /// running and accepting connection before starting netbench clients.
+    /// Should only be called by Coordinators.
+    WorkerRunning,
+}
+
+/// An instance of a protocol with an established connection to its peer.
+struct ProtocolInstance<P: Protocol> {
+    pub addr: SocketAddr,
+    pub stream: TcpStream,
+    pub protocol: P,
+}
+
+/// A Russula Endpoint.
+///
+/// An Endpoint can be of type Coordinator or Worker. A Coordinator can
+/// be used to synchronize multiple workers across different hosts. A Worker
+/// communicates with a Coordinator to make progress.
+pub struct Endpoint<P: Protocol> {
```

Comment:
It looks like this top level Endpoint struct has to be pretty involved with the low level details of polling all of the little downstream things. It there a way that we could push all of the polling details out of this struct?

I'm thinking that one appraoch might be to leverage the inherent "event passing" notification that you get when data is delivered to a socket with tokio. E.g. currently we
1. receive data on a TCP Stream
2. wait until some top level thing manually calls poll to update the state.

Rather than waiting to manually update things, could you just spawn a tokio task for each TCP Stream so that the state is automatically updated whenever you receive data in the socket? My hope is that this would make polling/updating the overall state significantly easier.